File: src/merkle/verifyMerkleProof.ts
import demosMemory from "./memory";

import libdemos from "@libdemos";

import { crypto_hash_sha512_BYTES } from "../utils/interfaces";

/**
 * Verifies that the hash was indeed included in the calculation of the Merkle root.
 * @param hash: The hash of the base element in question.
 * @param root: The Merkle root.
 * @param proof: The first element is the first leave that was added for the calculation etc. The last
 * byte is either 0 or 1, indicating whether it is to the left or to the right in the tree.
 */
const verifyMerkleProof = async (
  hash: Uint8Array,
  root: Uint8Array,
  proof: Uint8Array,
): Promise<boolean> => {
  const proofLen = proof.length;
  if (proofLen % (crypto_hash_sha512_BYTES + 1) !== 0)
    throw new Error("Proof length not multiple of hash length + 1.");

  const proofArtifactsLen = proofLen / (crypto_hash_sha512_BYTES + 1);

  const wasmMemory = demosMemory.verifyMerkleProofMemory(proofLen);
  const demosModule = await libdemos({
    wasmMemory,
  });

  const ptr1 = demosModule._malloc(crypto_hash_sha512_BYTES);
  const elementHash = new Uint8Array(
    wasmMemory.buffer,
    ptr1,
    crypto_hash_sha512_BYTES,
  );
  elementHash.set(hash);

  const ptr2 = demosModule._malloc(crypto_hash_sha512_BYTES);
  const rootArray = new Uint8Array(
    wasmMemory.buffer,
    ptr2,
    crypto_hash_sha512_BYTES,
  );
  rootArray.set(root);

  const ptr3 = demosModule._malloc(proofLen);
  const proofArray = new Uint8Array(wasmMemory.buffer, ptr3, proofLen);
  proofArray.set(proof);

  const result = demosModule._verify_merkle_proof(
    proofArtifactsLen,
    elementHash.byteOffset,
    rootArray.byteOffset,
    proofArray.byteOffset,
  );

  demosModule._free(ptr1);
  demosModule._free(ptr2);
  demosModule._free(ptr3);

  switch (result) {
    case 0:
      return true;

    case 1:
      return false;

    case -1:
      throw new Error(
        "Could not allocate memory for hash concatenation helper array.",
      );

    case -2:
      throw new Error("Could not allocate memory for hashes helper array.");

    case -3:
      throw new Error("Proof artifact position is neither left nor right.");

    case -4:
      throw new Error("Could not calculate hash.");

    default:
      throw new Error("Unexpected error occured.");
  }
};

export default verifyMerkleProof;

File: src/merkle/verify_merkle_proof.c
#include <stdbool.h>
#include <stdint.h>
#include <string.h>

#include "../../libsodium/src/libsodium/include/sodium/crypto_hash_sha512.h"

int
verify_merkle_proof(
    const unsigned int PROOF_ARTIFACTS_LEN,
    const uint8_t element_hash[crypto_hash_sha512_BYTES],
    const uint8_t root[crypto_hash_sha512_BYTES],
    const uint8_t proof[PROOF_ARTIFACTS_LEN * (crypto_hash_sha512_BYTES + 1)])
{
  int res;
  size_t i, position;

  if (PROOF_ARTIFACTS_LEN == 1)
  {
    res = memcmp(root, element_hash, crypto_hash_sha512_BYTES);
    if (res == 0)
    {
      return 0;
    }
    else
    {
      // If 1 artifact then either proof is the same
      // as the element hash, and therefore
      // the same as the root, or we have an error
      return -1;
    }
  }

  uint8_t(*concat_hashes)[crypto_hash_sha512_BYTES]
      = malloc(sizeof(uint8_t[2][crypto_hash_sha512_BYTES]));
  if (concat_hashes == NULL) return -1;

  uint8_t *hash = (uint8_t *)malloc(sizeof(uint8_t[crypto_hash_sha512_BYTES]));
  if (hash == NULL)
  {
    free(concat_hashes);

    return -2;
  }

  memcpy(hash, element_hash, crypto_hash_sha512_BYTES);

  /* for (i = 0; i < NODES_LEN; i++) */
  for (i = 0; i < PROOF_ARTIFACTS_LEN; i++)
  {
    position
        = proof[i * (crypto_hash_sha512_BYTES + 1) + crypto_hash_sha512_BYTES];
    if (position != 0 && position != 1)
    {
      free(hash);
      free(concat_hashes);

      return -3;
    }

    // Proof artifact goes to the left
    if (position == 0)
    {
      memcpy(concat_hashes[0], &proof[i * (crypto_hash_sha512_BYTES + 1)],
             crypto_hash_sha512_BYTES);
      memcpy(concat_hashes[1], hash, crypto_hash_sha512_BYTES);
    }
    else
    {
      memcpy(concat_hashes[0], hash, crypto_hash_sha512_BYTES);
      memcpy(concat_hashes[1], &proof[i * (crypto_hash_sha512_BYTES + 1)],
             crypto_hash_sha512_BYTES);
    }

    res = crypto_hash_sha512(hash, concat_hashes[0],
                             2 * crypto_hash_sha512_BYTES);
    if (res != 0)
    {
      free(concat_hashes);
      free(hash);

      return -4;
    }
  }

  free(concat_hashes);
  res = memcmp(hash, root, crypto_hash_sha512_BYTES);
  free(hash);

  if (res != 0) return 1;

  return 0;
}
